docs/specs/app2app.md

```diff
@@ -56,12 +56,16 @@ An app can start the authentication flow by opening a link to another app, inste
 
   - For `grant_type=authorization_code`, a new parameter `x_app2app_device_key_jwt` is supported. This value can be specified by the client to enable app2app login. If specified, it should be a jwt signed by the private key of the device key, and containing the public key of that key pair in header, with a `challenge` obtained from server in payload. `x_app2app_device_key_jwt` will be ignored if `x_app2app_enabled` of the client configuration is not `true`.
 
-  - A new grant_type `urn:authgear:params:oauth:grant-type:app2app` is supported. When using such grant_type, The following must be provided in the request:
-    - `refresh_token`: a valid refresh token.
-    - `client_id`: current client id.
-    - `app2app_client_id`: the authenticating client id.
-    - `jwt`: a jwt with a challenge token obtained from the `/oauth2/challenge` api, signed with the private key of the device key bound to the refresh token. If `x_app2app_insecure_device_key_binding_enabled` is `true`, a jwt with a new device key will be allowed if there were no device key bound in the current offline grant. In this case the new key will be bound to the offline grant after the request.
-    - `app2app_redirect_uri`: the redirect uri used to return the result used by the app. The server is reponsible to check the provided uri is in the whitelist of at least one of the configured client redirect uri.
+  - A new grant_type `urn:authgear:params:oauth:grant-type:app2app` is supported. When using such grant_type, The following request parameters are supported:
+    - `refresh_token`: Required. A valid refresh token.
+    - `client_id`: Required. Current client id.
+    - `app2app_client_id`: Required. The authenticating client id.
+    - `jwt`: Required. A jwt with a challenge token obtained from the `/oauth2/challenge` api, signed with the private key of the device key bound to the refresh token. If `x_app2app_insecure_device_key_binding_enabled` is `true`, a jwt with a new device key will be allowed if there were no device key bound in the current offline grant. In this case the new key will be bound to the offline grant after the request.
+    - `app2app_redirect_uri`: Required. The redirect uri used to return the result used by the app. The server is reponsible to check the provided uri is in the whitelist of at least one of the configured client redirect uri.
+    - `login_hint`: Optional. When provided, it must be in the format described in [here](./oidc.md#login_hint). The server will check the corresponding user identity referenced by the `refresh_token`, and if it does not match the provided `login_hint`, a `unmatched_user` error will be returned. Only the following combinations of parameters are supported.
```

Comment:
In the client's use case, App B actually can perform the checking locally by itself.

When the end-user has authenticated in App B
- App B is opened by the authorization URL. It see login_hint in the the query parameters. It then call the userinfo endpoint to fetch the standard attributes.
- If the phone_number does not match, App B return directly to App A.
- If the phone_number match, App B proceeds.

When the end-user has NOT authenticated in App B
- App B construct an authorization URL to include the login_hint.
- The end-user authenticated.
- App B calls the userinfo endpoint to check the login_hint.

The reason I am proposing this is because login_hint is originally a hint to the authorization server. In our original spec, we explicitly states that the login_hint will be ignored when it cannot be fulfilled. But in here, we are changing its meaning to be mandatory.

Reply:
I am trying to support login_hint in server is to ensure the checking logic is same in all sdks. But i agree that we should not change the meaning of login_hint at server. Let me  remove the parameter from token endpoint.

Reply:
Per our offline discussion, instead of introducing login_hint, we will support `state` in app2app request. Its meaning is the same as in the authorization code flow. It will be included in the redirect URI to App A.

Reply:
Updated to add state only in the spec, thanks!